### PR TITLE
1750 start date and throw scmdata back

### PR DESCRIPTION
**Pull request by chrisroadmap**
Flexibility to allow FaIR to run from a non-1765 start date. Implemented because forcing usually reported from 1750.

scmdata to FaIR emissions files converter taken out of FaIR and put into openscm-runner.

- [x] Documentation added
- [x] Description in ``CHANGELOG.rst`` added (single line such as: ``(`#XX <https://github.com/openscm/openscm-runner/pull/XX>`_) Added feature which does something``)

### CODE REVIEW

**Comment by chrisroadmap** [comment]
Less of a flexibility - it's still hardcoded, just with a 1750 start date instead of 1765. Not sure how to get this flexibility into a keyword option in the current interface to the adapter. It might be useful to use an 1850 start date for example for CMIP-like runs.

File: CHANGELOG.rst [hunk 1]
```diff
@@ -19,6 +19,8 @@ The changes listed in this file are categorised as follows:
 master
 ------
 
+- (`#21 <https://github.com/openscm/openscm-runner/pull/21>`_) Flexible start date for FaIR; build scmdata to emissions converter on openscm-runner
```
Comment (znicholls) [inline CHANGELOG.rst:22]:
```suggestion
- (`#21 <https://github.com/openscm/openscm-runner/pull/21>`_) Added flexible start date for FaIR and FaIR's scmdata to emissions converter
```

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
+"""
+Converter for reading scmdata dataframes into FaIR emissions arrays.
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:2]:
```suggestion
Conversion of :obj:`scmdata.ScmRun` into FaIR emissions :obj:`np.ndarray`.
```

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    def __init__(self):
+        self._loaded = False
+        self._loaded_fair_history = False
+        self._values = None
+        self._values_fair_history = None
+
+    @property
+    def values(self):
+        """Fill in emissions values from historical + SSP2-4.5"""
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:23]:
```suggestion
        """Emissions values from historical joint with ssp245"""
```

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            self._values = self._values.interpolate(
+                [dt.datetime(y, 1, 1) for y in self._values["year"]]
+            )
+
+        self._loaded = True
+        return self._values
+
+    @property
+    def values_fair_units(self):
+        """Fill in the units and ignores species not used by FaIR"""
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:41]:
```suggestion
        """Get values with FaIR's expected units, ignores species not used by FaIR"""
```

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                    )
+
+            self._values_fair_history = ssp_df_hist
+
+        self._loaded_fair_history = True
+
+        return self._values_fair_history
+
+
+# TODO: lazy load this and only load once
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:82]:
```suggestion
```

I think it already is lazy loaded

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    context = EMISSIONS_SPECIES_UNITS_CONTEXT[row]["context"].iloc[0]
+
+    return fair_col, in_unit, context
+
+
+def scmdf_to_emissions(
+    scmrun, startyear=1750, endyear=2100, scen_startyear=2015
+):  # pylint: disable=R0914
+    """
+    Open an :obj:`ScmRun` and extracts the data. Interpolates linearly
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:152]:
```suggestion
    Convert an :obj:`scmdata.ScmRun` into a FaIR emissions :obj:`np.ndarray`

    Interpolates linearly if required and fills in montreal gases based on ssp245.
```

File: src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    ------
+    AssertionError
+        If there is more than one model-scenario pair in the provided :obj:`ScmRun`.
+    """
+    # We expect that aeneris and silicone are going to give us a nicely
+    # formatted ScmRun with all 23 species present and correct at
+    # timesteps 2015, 2020 and ten-yearly to 2100.
+    # We also implicitly assume that data up until 2014 will follow SSP
+    # historical.
+    # This adapter will not be tested on anything else!
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/_scmdf_to_emissions.py:182]: Can probably delete this comment

File: tests/integration/test_fair1X.py [hunk 8]
```diff
@@ -49,95 +48,19 @@ def test_fair_run(test_scenarios):
...
-    npt.assert_allclose(
-        4.07184261082167,
-        res.filter(
-            variable="Surface Temperature",
-            region="World",
-            year=2100,
-            scenario="ssp370",
-        ).values.min(),
-        rtol=RTOL,
+    res.filter(
```
Comment (znicholls) [inline tests/integration/test_fair1X.py:51]: we want to have at least some tests of values in here...

**Comment by chrisroadmap** [comment]
@znicholls I now have some csv files in fair_adapter - is this the best place for them?

**Comment by znicholls** [comment]
> is this the best place for them?

yep

**Comment by chrisroadmap** [comment]
OK @znicholls ready. Shame 1750 is hardcoded, but this is more useful than 1765. Changing it I think would need an extra parameter to the runner function.

File: src/openscm_runner/adapters/fair_adapter/fair_adapter.py [hunk 9]
```diff
@@ -38,36 +40,40 @@ def _make_full_cfgs(self, scenarios, cfgs):  # pylint: disable=R0201
             desc="Creating FaIR emissions inputs",
         ):
             smdf_in = ScmRun(smdf)
+            scen_startyear = smdf_in.time_points.years()[0]
             endyear = smdf_in.time_points.years()[-1]
-            emissions = scmdf_to_emissions(smdf_in, endyear=endyear)
+            emissions = scmdf_to_emissions(
```
Comment (znicholls) [inline src/openscm_runner/adapters/fair_adapter/fair_adapter.py:45]: If you wanted to make the start year flexible, you could do something like

```
startyear = cfgs.pop("startyear", 1750)
scmdf_to_emissions(startyear=startyear)
```

That would let a user tell FaIR exactly which year to start in without having that config then propagate everywhere (I think)

**Comment by znicholls** [comment]
Looks good to me. If you want to fix that startyear thing, go for it. Otherwise we can merge and leave for another day

**Comment by chrisroadmap** [comment]
Rebased on to master

**Comment by znicholls** [comment]
> Rebased on to master

Assuming you don't want to fix the startyear?

**Comment by chrisroadmap** [comment]
> > Rebased on to master
> 
> Assuming you don't want to fix the startyear?

Yep doing this now, making sure it doesn't break everything etc

**Comment by znicholls** [comment]
Beauty @chrisroadmap, will release new version now

**Pull request merged**